File: docs/testing/generate-test-report-excel.py
#!/usr/bin/env python3
"""Xuat bao-cao-kiem-thu.xlsx (fallback khi PHP khong co ext-zip)."""
from __future__ import annotations

import json
import subprocess
from pathlib import Path

from openpyxl import Workbook
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side
from openpyxl.utils import get_column_letter

ROOT = Path(__file__).resolve().parents[2]
PHP = Path(r"C:\xampp\php\php.exe")
DATA_PHP = ROOT / "docs" / "testing" / "test-report-data.php"
OUTPUT = ROOT / "docs" / "testing" / "bao-cao-kiem-thu.xlsx"

HEADER_FILL = PatternFill("solid", fgColor="1E3A5F")
HEADER_FONT = Font(name="Times New Roman", bold=True, color="FFFFFF", size=11)
BODY_FONT = Font(name="Times New Roman", size=11)
PASS_FILL = PatternFill("solid", fgColor="D4EDDA")
THIN = Side(style="thin", color="999999")
BORDER = Border(left=THIN, right=THIN, top=THIN, bottom=THIN)


def load_data() -> dict:
    cmd = [str(PHP), "-r", f"echo json_encode(require '{DATA_PHP.as_posix()}', JSON_UNESCAPED_UNICODE);"]
    raw = subprocess.check_output(cmd, cwd=ROOT, text=True, encoding="utf-8")
    return json.loads(raw)


def style_header_row(ws, row: int, cols: int) -> None:
    for c in range(1, cols + 1):
        cell = ws.cell(row=row, column=c)
        cell.fill = HEADER_FILL
        cell.font = HEADER_FONT
        cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)
        cell.border = BORDER


def write_table(ws, start_row: int, headers: list[str], rows: list[list]) -> int:
    for c, h in enumerate(headers, 1):
        ws.cell(row=start_row, column=c, value=h)
    style_header_row(ws, start_row, len(headers))
    r = start_row + 1
    for row in rows:
        for c, val in enumerate(row, 1):
            cell = ws.cell(row=r, column=c, value=val)
            cell.font = BODY_FONT
            cell.alignment = Alignment(vertical="top", wrap_text=True)
            cell.border = BORDER
            if headers[c - 1] == "Kết quả" and val == "Pass":
                cell.fill = PASS_FILL
        r += 1
    return r


def autosize(ws, widths: list[int]) -> None:
    for i, w in enumerate(widths, 1):
        ws.column_dimensions[get_column_letter(i)].width = w


def main() -> None:
    data = load_data()
    meta = data["meta"]
    sections = data["sections"]

    total = pass_count = fail_count = 0
    for section in sections:
        for case in section["cases"]:
            total += 1
            if case[4] == "Pass":
                pass_count += 1
            elif case[4] == "Fail":
                fail_count += 1
    rate = round(pass_count / total * 100, 1) if total else 0

    wb = Workbook()
    ws0 = wb.active
    ws0.title = "Tong quan"
    cover = [
        ["Bao cao kiem thu he thong", ""],
        [meta["project"], ""],
        [meta["subtitle"], ""],
        ["", ""],
        ["Phien ban", meta["version"]],
        ["Ngay kiem thu", meta["date"]],
        ["Nguoi thuc hien", meta["tester"]],
        ["Phuong phap", meta["method"]],
        ["", ""],
        ["Tong TC", "Pass", "Fail", "Ty le (%)"],
        [total, pass_count, fail_count, rate],
    ]
    for r, row in enumerate(cover, 1):
        for c, val in enumerate(row, 1):
            ws0.cell(row=r, column=c, value=val).font = BODY_FONT
    autosize(ws0, [28, 18, 12, 12])

    ws1 = wb.create_sheet("Tong hop module")
    mod_rows = []
    for i, s in enumerate(sections, 1):
        st = len(s["cases"])
        sp = sum(1 for c in s["cases"] if c[4] == "Pass")
        mod_rows.append([i, s["id"], s["title"], st, sp, st - sp, round(sp / st * 100, 1) if st else 0])
    mod_rows.append(["", "Tong", "", total, pass_count, fail_count, rate])
    write_table(ws1, 1, ["STT", "Module", "Ten module", "So TC", "Pass", "Fail", "Ty le (%)"], mod_rows)
    autosize(ws1, [6, 10, 42, 10, 10, 10, 12])

    ws2 = wb.create_sheet("Chi tiet test case")
    detail = []
    seq = 1
    for s in sections:
        for case in s["cases"]:
            detail.append([seq, case[0], s["id"], s["title"], case[1], case[2], case[3], case[4], case[5], "", "", ""])
            seq += 1
    write_table(
        ws2,
        1,
        ["STT", "Ma TC", "Module", "Ten module", "Chuc nang", "Cac buoc", "Ket qua mong doi", "Ket qua", "Uu tien", "Nguoi test", "Ngay test", "Ghi chu"],
        detail,
    )
    autosize(ws2, [5, 11, 8, 28, 22, 36, 38, 10, 12, 14, 12, 20])

    for s in sections:
        name = f"Module {s['id']}"[:31]
        ws = wb.create_sheet(name)
        rows = [[c[0], c[1], c[2], c[3], c[4], c[5]] for c in s["cases"]]
        write_table(ws, 1, ["Ma TC", "Chuc nang", "Cac buoc", "Ket qua mong doi", "Ket qua", "Uu tien"], rows)
        autosize(ws, [11, 24, 36, 40, 10, 12])

    wb.save(OUTPUT)
    print(str(OUTPUT))


if __name__ == "__main__":
    main()

File: docs/testing/generate-test-report-excel.php
<?php

declare(strict_types=1);

/**
 * Sinh báo cáo kiểm thử Excel (.xlsx) có định dạng cho đồ án Winsum Home.
 * Chạy: C:\xampp\php\php.exe docs/generate-test-report-excel.php
 */

$docsDir = __DIR__;
require_once dirname($docsDir) . DIRECTORY_SEPARATOR . 'includes' . DIRECTORY_SEPARATOR . 'styled-xlsx-writer.php';

$data = require $docsDir . DIRECTORY_SEPARATOR . 'test-report-data.php';
$meta = $data['meta'];
$sections = $data['sections'];
$outputPath = $docsDir . DIRECTORY_SEPARATOR . 'bao-cao-kiem-thu.xlsx';

$totalCases = 0;
$passCount = 0;
$failCount = 0;
foreach ($sections as $section) {
    foreach ($section['cases'] as $case) {
        $totalCases++;
        if (($case[4] ?? '') === 'Pass') {
            $passCount++;
        } elseif (($case[4] ?? '') === 'Fail') {
            $failCount++;
        }
    }
}
$passRate = $totalCases > 0 ? round(($passCount / $totalCases) * 100, 1) : 0;

$writer = new StyledXlsxWriter();

// --- Sheet 1: Tổng quan (bìa) ---
$cover = [];
$blank = static fn() => ['value' => '', 'style' => StyledXlsxWriter::STYLE_DEFAULT];
$cell = static fn($v, int $style) => ['value' => $v, 'style' => $style];

$cover[] = [$blank(), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('BÁO CÁO KIỂM THỬ HỆ THỐNG', StyledXlsxWriter::STYLE_TITLE), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell($meta['project'], StyledXlsxWriter::STYLE_SUBTITLE), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell($meta['subtitle'], StyledXlsxWriter::STYLE_BODY), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$blank(), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('Thông tin chung', StyledXlsxWriter::STYLE_LABEL), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('Phiên bản', StyledXlsxWriter::STYLE_LABEL), $cell($meta['version'], StyledXlsxWriter::STYLE_BODY), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('Ngày kiểm thử', StyledXlsxWriter::STYLE_LABEL), $cell($meta['date'], StyledXlsxWriter::STYLE_BODY), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('Người thực hiện', StyledXlsxWriter::STYLE_LABEL), $cell($meta['tester'], StyledXlsxWriter::STYLE_BODY), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('Phương pháp', StyledXlsxWriter::STYLE_LABEL), $cell($meta['method'], StyledXlsxWriter::STYLE_BODY), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$blank(), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('Tổng hợp kết quả', StyledXlsxWriter::STYLE_LABEL), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [
    $cell('Tổng test case', StyledXlsxWriter::STYLE_HEADER),
    $cell('Pass', StyledXlsxWriter::STYLE_HEADER),
    $cell('Fail', StyledXlsxWriter::STYLE_HEADER),
    $cell('Tỷ lệ đạt (%)', StyledXlsxWriter::STYLE_HEADER),
    $blank(),
    $blank(),
];
$cover[] = [
    $cell($totalCases, StyledXlsxWriter::STYLE_BODY),
    $cell($passCount, StyledXlsxWriter::STYLE_PASS),
    $cell($failCount, StyledXlsxWriter::STYLE_FAIL),
    $cell($passRate, StyledXlsxWriter::STYLE_BODY),
    $blank(),
    $blank(),
];
$cover[] = [$blank(), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('Môi trường kiểm thử', StyledXlsxWriter::STYLE_LABEL), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('Thành phần', StyledXlsxWriter::STYLE_HEADER), $cell('Cấu hình', StyledXlsxWriter::STYLE_HEADER), $blank(), $blank(), $blank(), $blank()];
foreach ($meta['environment'] as $k => $v) {
    $cover[] = [$cell($k, StyledXlsxWriter::STYLE_LABEL), $cell($v, StyledXlsxWriter::STYLE_BODY), $blank(), $blank(), $blank(), $blank()];
}
$cover[] = [$blank(), $blank(), $blank(), $blank(), $blank(), $blank()];
$cover[] = [$cell('Kết luận', StyledXlsxWriter::STYLE_LABEL), $blank(), $blank(), $blank(), $blank(), $blank()];
$conclusionRow = count($cover) + 1;
$cover[] = [$cell($meta['conclusion'], StyledXlsxWriter::STYLE_BODY), $blank(), $blank(), $blank(), $blank(), $blank()];

$writer->addRawSheet(
    'Tổng quan',
    $cover,
    ['A2:F2', 'A3:F3', 'A4:F4', 'A' . $conclusionRow . ':F' . $conclusionRow],
    0,
    [22, 28, 12, 12, 12, 12]
);

// --- Sheet 2: Tổng hợp module ---
$moduleRows = [['STT', 'Module', 'Tên module', 'Số TC', 'Pass', 'Fail', 'Tỷ lệ (%)']];
$stt = 1;
foreach ($sections as $section) {
    $sTotal = count($section['cases']);
    $sPass = count(array_filter($section['cases'], static fn($c) => ($c[4] ?? '') === 'Pass'));
    $sFail = $sTotal - $sPass;
    $rate = $sTotal > 0 ? round(($sPass / $sTotal) * 100, 1) : 0;
    $moduleRows[] = [$stt++, $section['id'], $section['title'], $sTotal, $sPass, $sFail, $rate];
}
$moduleRows[] = ['', 'Tổng', '', $totalCases, $passCount, $failCount, $passRate];
$moduleLastRow = count($moduleRows) - 1;

$writer->addSheet(
    'Tổng hợp module',
    $moduleRows,
    [],
    1,
    [6, 10, 42, 10, 10, 10, 12],
    'A1:G' . count($moduleRows),
    static function ($value, int $r, int $c) use ($moduleLastRow): int {
        if ($r === 0) {
            return StyledXlsxWriter::STYLE_HEADER;
        }
        if ($r === $moduleLastRow) {
            return StyledXlsxWriter::STYLE_TOTAL;
        }
        if ($c === 4) {
            return StyledXlsxWriter::STYLE_PASS;
        }
        if ($c === 5 && (int) $value > 0) {
            return StyledXlsxWriter::STYLE_FAIL;
        }
        return StyledXlsxWriter::STYLE_BODY;
    }
);

// --- Sheet 3: Chi tiết test case ---
$detailHeader = ['STT', 'Mã TC', 'Module', 'Tên module', 'Chức năng', 'Các bước thực hiện', 'Kết quả mong đợi', 'Kết quả', 'Ưu tiên', 'Người test', 'Ngày test', 'Ghi chú'];
$detailRows = [$detailHeader];
$seq = 1;
foreach ($sections as $section) {
    foreach ($section['cases'] as $case) {
        $detailRows[] = [
            $seq++,
            $case[0],
            $section['id'],
            $section['title'],
            $case[1],
            $case[2],
            $case[3],
            $case[4],
            $case[5] ?? '',
            '',
            '',
            '',
        ];
    }
}

$writer->addSheet(
    'Chi tiết test case',
    $detailRows,
    [],
    1,
    [5, 11, 8, 28, 22, 36, 38, 10, 12, 14, 12, 20],
    'A1:L' . count($detailRows),
    static function ($value, int $r, int $c): int {
        if ($r === 0) {
            return StyledXlsxWriter::STYLE_HEADER;
        }
        if ($c === 7) {
            if ($value === 'Pass') {
                return StyledXlsxWriter::STYLE_PASS;
            }
            if ($value === 'Fail') {
                return StyledXlsxWriter::STYLE_FAIL;
            }
        }
        return StyledXlsxWriter::STYLE_BODY;
    }
);

// --- Sheet: Hạn chế & hướng phát triển ---
$notes = [
    ['Nội dung', 'Chi tiết'],
    ['Hạn chế kiểm thử', ''],
];
foreach ($meta['limitations'] as $item) {
    $notes[] = ['', '• ' . $item];
}
$notes[] = ['', ''];
$notes[] = ['Hướng phát triển', ''];
foreach ($meta['future'] as $item) {
    $notes[] = ['', '• ' . $item];
}

$writer->addSheet(
    'Hạn chế & phát triển',
    $notes,
    [],
    1,
    [24, 70],
    null,
    static function ($value, int $r, int $c): int {
        if ($r === 0) {
            return StyledXlsxWriter::STYLE_HEADER;
        }
        if ($c === 0 && $value !== '') {
            return StyledXlsxWriter::STYLE_LABEL;
        }
        return StyledXlsxWriter::STYLE_BODY;
    }
);

// --- Sheet từng module ---
foreach ($sections as $section) {
    $rows = [['Mã TC', 'Chức năng', 'Các bước', 'Kết quả mong đợi', 'Kết quả', 'Ưu tiên']];
    foreach ($section['cases'] as $case) {
        $rows[] = [$case[0], $case[1], $case[2], $case[3], $case[4], $case[5] ?? ''];
    }
    $writer->addSheet(
        'Module ' . $section['id'],
        $rows,
        [],
        1,
        [11, 24, 36, 40, 10, 12],
        'A1:F' . count($rows),
        static function ($value, int $r, int $c): int {
            if ($r === 0) {
                return StyledXlsxWriter::STYLE_HEADER;
            }
            if ($c === 4) {
                if ($value === 'Pass') {
                    return StyledXlsxWriter::STYLE_PASS;
                }
                if ($value === 'Fail') {
                    return StyledXlsxWriter::STYLE_FAIL;
                }
            }
            return StyledXlsxWriter::STYLE_BODY;
        }
    );
}

try {
    $writer->save($outputPath);
    $sizeKb = round(filesize($outputPath) / 1024, 1);
    echo "Đã tạo Excel: {$outputPath} ({$sizeKb} KB)\n";
    echo "Tổng: {$totalCases} TC | Pass: {$passCount} | Fail: {$failCount} | Tỷ lệ: {$passRate}%\n";
} catch (Throwable $e) {
    fwrite(STDERR, 'Lỗi: ' . $e->getMessage() . "\n");
    exit(1);
}
